core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantDistinctAggregation.java

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Lookup;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+
+/**
+ * Removes DISTINCT only aggregation, when the input source is already distinct over a subset of
+ * the grouping keys as a result of another aggregation.
+ *
+ * Given:
+ * <pre>
+ * - Aggregate[keys = [a, max]]
+ *   - Aggregate[keys = [a]]
+ *     max := max(b)
+ * </pre>
+ * <p>
+ * Produces:
+ * <pre>
+ *   - Aggregate[keys = [a]]
+ *     max := max(b)
+ * </pre>
+ */
+public class RemoveRedundantDistinctAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(RemoveRedundantDistinctAggregation::isDistinctOnlyAggregation);
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        if (isDistinctOverGroupingKeys(lookup.resolve(aggregationNode.getSource()), lookup, new HashSet<>(aggregationNode.getGroupingKeys()))) {
+            return Result.ofPlanNode(aggregationNode.getSource());
+        }
+        else {
+            return Result.empty();
+        }
+    }
+
+    private static boolean isDistinctOnlyAggregation(AggregationNode node)
+    {
+        return node.producesDistinctRows() && node.getGroupingSetCount() == 1;
+    }
+
+    private static boolean isDistinctOverGroupingKeys(PlanNode node, Lookup lookup, Set<Symbol> groupingKeys)
+    {
+        return switch (node) {
+            case AggregationNode aggregationNode -> aggregationNode.getGroupingSets().getGroupingSetCount() == 1 &&
+                    groupingKeys.containsAll(aggregationNode.getGroupingSets().getGroupingKeys());
+
+            // Project nodes introduce new symbols for computed expressions, and therefore end up preserving distinctness
+            // between the distinct aggregation and the child aggregation nodes so long as all child aggregation keys
+            // remain present (without transformation by the project) in the distinct aggregation grouping keys
+            case ProjectNode projectNode ->
+                    projectNode.isIdentity() && isDistinctOverGroupingKeys(lookup.resolve(projectNode.getSource()), lookup, groupingKeys);
```

Comment:
`projectNode.isIdentity()` is too restrictive here- there can be new symbols introduced that are non-identity as long as the symbols from the child grouping keys are preserved unmodified. That actually makes this kind of awkward to test during recursion since we actually need to know the child grouping keys.

@martint - I thought that any non-identity transformations would result in a new symbol which would indirectly guarantee this property. Is that wrong? Because if so, we would first need to recurse down to find the child aggregation and then pass the child grouping keys back up the call stack to verify those.

Reply:
> I thought that any non-identity transformations would result in a new symbol which would indirectly guarantee this property

That's generally the case, but there are no guarantees. Symbols should be thought of as "local variables" (the output of an operation gets assigned a name so that it can be referenced in another expression). There are some inconsistencies, though -- for instance, in some places we keep global maps of symbol -> type, so there's a implementation constraint that if a symbol is reused, its type should not change. This is all part of a much larger conversation on what the next generation of the IR should look like (hint: I want to get rid of symbols in their current form)

Reply:
Makes sense, then we'll have to update the logic to pass the child aggregation's grouping key symbols back up the call stack and check them against the `ProjectNode` to make sure that all of those are identity projections before proceeding with the optimization.

Reply:
I have updated the logic as discussed here, thank you.